### PR TITLE
manage projects via api

### LINKED ISSUE #124: update projects from data bag items
currently the cookbook will only create new projects from data bags, but wont update existing project config via chef-client. add support for updating config of existing projects via chef-client.

this probably would be covered by a project lwrp #24

### LINKED ISSUE #136: Project creation does not happen on first chef-client
**tl;dr: Turns out rundeck isnt running yet when projects are trying to be created. The project creation logic needs to be moved out of rundeck::server and into a new recipe after rundeck::apache is run.**

Projects are currently not created on the first chef-client run (in test-kitchen at least, Im not sure about with a real node). ~~I think there is an issue with apache or rundeckd restarting too close to right before the `rd-project -p #{project} -a create` command is run in the server_install recipe.~~ edit: i debunked this below

also, it doesnt help that the project creation is guarded by
```ruby
    # will return 0 if grep matches
    # only run if project does not exist
    only_if "rd-jobs -p #{project} list 2>&1 | grep -q '^ERROR .*project does not exist'"
```

I assume `rd-jobs` is returning some other error (`rundeckd not running` perhaps) and project creation is being skipped.

I identified this bug by adding a localhost port forward from my mac into the vm (`8080 on host => 80 on guest`), logging in as the admin, and seeing no projects. After converging the node again, I login and see the localhost project. You can also see in the chef-client output from the initial run:

```
         * bash[check-project-localhost] action run (skipped due to only_if)
```

and from the second run

```
         * bash[check-project-localhost] action run
           - execute "bash"  "/tmp/chef-script20170128-5488-f89mr3"
```

When projects are managed via the api client, this should be resolved I think. If not, the issue will become more obvious (the api will respond with a 500).

### PR DESCRIPTION
**Pull request by atheiman**
not yet ready for review

fixes #124 

- [x] version change
- [x] README updates
- [x] project creation happening on first chef-client (fixes #136)
- [x] project lwrp
  - [x] create ([POST with config](http://rundeck.org/docs/api/#project-creation))
  - [x] always (create and update) ([PUT all config](http://rundeck.org/docs/api/#put-project-configuration))
- [x] create api client into `node.run_state`
- [x] update recipes to use lwrp
- [x] update kitchen tests to verify projects created and updated as expected
- [x] update chefspec tests to verify projects created and updated as expected
- [x] specify extra config in framework.properties with `['rundeck']['framework']['properties']`

### CODE REVIEW

File: libraries/rundeck_helper.rb [hunk 1]
```diff
@@ -6,4 +6,33 @@ module RundeckHelper
   def self.generateuuid
     SecureRandom.uuid
   end
+
+  def secure_databag
```
Comment (atheiman) [inline libraries/rundeck_helper.rb]: ill probably move this out to a separate pr to be merged before this one

**Comment by atheiman** [comment]
i implemented a lot of this in #135 ill rebase on that once its merged

File: recipes/server.rb [hunk 2]
```diff
@@ -17,6 +17,23 @@
...
 #
 
+node.run_state['rundeck'] = Hash.recursive
+
+ruby_block "connect rundeck api client" do
+  action :nothing
+  block do
+    unless node.run_state['rundeck']['api_client']
+      admin_user = node['rundeck']['admin_user']
+      node.run_state['rundeck_api_client'] = RundeckApiClient.connect(
```
Comment (atheiman) [inline recipes/server.rb]: `['rundeck']['api_client']`

**Comment by atheiman** [comment]
ive pinned back kitchen-dokken version because I was running into https://github.com/someara/kitchen-dokken/issues/62

**Comment by atheiman** [comment]
also, the cookbook is [`chef-client`](https://supermarket.chef.io/cookbooks/chef-client), not `chef_client`

**Comment by edwlarkey** [comment]
Might be worth a try to bypass apache and hit rundeck directly at `http://localhost:4440`

**Comment by atheiman** [comment]
the issue seems to be the project is created with the POST and then config is set with the PUT immediately. I think if I sleep for a few seconds between creating the project and updating the config things will work.

**Comment by atheiman** [comment]
one thing to note with this pr (at least in its current state) is that projects will no longer be created with certain config, then have additional config appended by Rundeck. Projects will now exist _exactly_ as they are defined in the project data bag items.

**Comment by atheiman** [comment]
welp so much for that idea, im occasionally getting 503s on the POST to create the project as well. It must just be an issue with the service not fully starting up before Im creating the projects.

**Comment by atheiman** [comment]
since Im bumping major version already for project data bag functionality change, Im going to get rid of the backwards compatibility data bag functionality that does this: https://github.com/Webtrends/rundeck/blob/3865dc95cc3da033a9346680991a5cc29376c2be/recipes/server_install.rb#L270-L277

the new data bags will have configuration exactly as the api expects to upload.

**Comment by atheiman** [comment]
i could add `converge_if_changed` to the project lwrp to output changes happening on the project config https://docs.chef.io/custom_resources.html#converge-if-changed

**Comment by atheiman** [comment]
@pcross616 @ARentz07 can you review this

**Comment by bpcook61** [comment]
+1

**Comment by edwlarkey** [comment]
+1
